fix(compiler-cli): ignoring useFactory as method declaration in Injectable

crisbeto opened the pull request:
Currently when we're analyzing an object literal that is passed into an Angular decorator, we only look for property assignments, which covers most cases, but breaks down if something is defined as a method declaration, e.g.

```ts
@Injectable({
  // Won't be picked up
  useFactory() {
    return new Service();
  }
})
class Service {}
```

These changes make it so that we pick up method declarations too.

In the process of working on this issue, I found another one: we were generating invalid code when emitting a call to a `FunctionExpression`. The problem was that we'd create a `CallExpression` directly from the `FunctionExpression` which would result an expression like `function foo() {}()`. I've added some code so that we wrap the function in parentheses: `(function foo() {})()`.

Fixes #38578.

**Note:** I didn't add any equivalent tests for the runtime, because the runtime tests are in JIT mode which doesn't have the problem.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -513,6 +513,16 @@ export function reflectObjectLiteral(node: ts.ObjectLiteralExpression): Map<stri
+      }
+      map.set(
+          name,
+          ts.createFunctionExpression(
```
Do we need to concern ourselves with the binding of `this` in such a method declaration?
For example, would such a methed ever be analysed by the partial evaluator? In which case the use of `this` inside the method would be affected if this function expression was treated as unbound.

petebacondarwin replied:
Would it work to store the method declaration directly rather than synthesizing a new function expression node, given that the translator does have logic to handle `visitInvokeMethodExpr()`?

crisbeto replied:
I don't think we have to concern ourselves with `this` since it's inside an object literal where it doesn't make sense. Also I don't think that we can store the method declaration directly, because:
1. Storing the method would produce different code, compared to a `FunctionExpression` (e.g. `useFactory() {}` vs `function useFactory() {}`).
2. The typings expect an `Expression`. It probably won't be too difficult to fix, but we'd have to update a lot of places.

petebacondarwin replied:
> I don't think we have to concern ourselves with this since it's inside an object literal where it doesn't make sense.

I am not sure I agree with this. It is certainly possible to refer to other fields in the object literal via `this` in such a method. Try running this [TS playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAHjAvDA3gKBjAhgLhgRgAYAaDGAMxBAAoBKVMzUSEAGwFMA6VkAc2qgALAJYROWWgG4yAXzRy0cTpRq0gA).

I agree that we can't store the method declaration directly. I think we would need to store a function that is bound to the appropriate `this`. I am not sure the best approach but two alternatives would be:

```
() => { obj.useFactory(); }
// where we would have to synthesize and pass through the parameters
```

or

```
useFactory.bind(obj)()
```

crisbeto replied:
How would we get a reference to `obj` though? We'd have to move the object into a variable that would be passed into `bind` and the `factory` property. I think that it complicates the compiler unnecessarily, considering that the only other property that is allowed on the object literal is `providedIn`.

petebacondarwin replied:
This function `reflectObjectLiteral()` is used for pretty much all of the decorators. But from a quick look through it seems like most if not all of the decorators do not allow functions as properties of the decorator metadata (except for this exception). Perhaps that is why this has not come up before?

So, OK at the moment I guess this would be adding unnecessary complexity to the compiler. But I am worried that it would be confusing for the developer if we do not follow the semantics of the code - perhaps in the future. I wonder if we could do an assertion that there are no references to `this` in the body of the function when reflecting on such a method before outputting it as a free standing function?

crisbeto replied:
Throwing an error if `this` is used inside the function would be easy, but I'm not completely sure where in the compilation pipeline would be best to throw the error (e.g. whether to do it directly in `reflectObjectLiteral` or somewhere further down the line). 

I'd have to play around with it, but maybe we can do it with types? TS allows for `this` to be typed in functions so we could do something like the following which will flag any accesses of `this`:


```
interface InjectableOptions {
    providedIn: string;
    useFactory?: (this: never) => any;
}

const service: InjectableOptions = {
    providedIn: 'root',
    useFactory() {
        return this.providedIn;
    }
};
```

petebacondarwin replied:
I think that typing the `InjectableOptions` interface is a great idea. Catches the problem really early on in the developer experience.

alxhub replied:
+1 for constraining the type of `useFactory` to infer `this: never` in these cases.

crisbeto replied:
Changed the type to `(this: never, ...args: any[]) => any`.

crisbeto committed:
fix(compiler-cli): ignoring useFactory as method declaration in Injectable

Currently when we're analyzing an object literal that is passed into an Angular decorator, we only
look for property assignments, which covers most cases, but breaks down if something is defined as
a method declaration, e.g.

```ts
@Injectable({
  // Won't be picked up
  useFactory() {
    return new Service();
  }
})
class Service {}
```

These changes make it so that we pick up method declarations too.

In the process of working on this issue, I found another one: we were generating invalid code when
emitting a call to a `FunctionExpression`. The problem was that we'd create a `CallExpression`
directly from the `FunctionExpression` which would result an expression like `function foo() {}()`.
I've added some code so that we wrap the function in parentheses: `(function foo() {})()`.

Fixes #38578.

crisbeto committed:
fixup! fix(compiler-cli): ignoring useFactory as method declaration in Injectable

crisbeto commented:
I've reworked it based on the discussion and accounted for some recent changes in the compiler. Can you take one more look at @petebacondarwin? Also added @jelbourn since I had to change a couple of public API signatures (`useFactory` in `Injectable` and `TestBed`), based on the discussion above.

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

petebacondarwin commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -513,6 +513,16 @@ export function reflectObjectLiteral(node: ts.ObjectLiteralExpression): Map<stri
       map.set(name, prop.initializer);
     } else if (ts.isShorthandPropertyAssignment(prop)) {
       map.set(prop.name.text, prop.name);
+    } else if (ts.isMethodDeclaration(prop) && prop.body !== undefined) {
+      const name = propertyNameToString(prop.name);
+      if (name === null) {
+        return;
+      }
+      map.set(
+          name,
+          ts.createFunctionExpression(
+              undefined, prop.asteriskToken, name, prop.typeParameters, prop.parameters, prop.type,
+              prop.body));
```
This is the first time we create TS nodes in a `ReflectionHost`. Generally the host is "readonly" and I am worried this might have repercussions later on.  But perhaps this is OK...
It is a shame that we cannot just add the `ts.MethodDeclaration` to the map directly and then let the translator worry about it later.

alxhub replied:
I agree, the `ReflectionHost` should not be doing this kind of operation. It should return "ground truth" data about the source, and the consumer should perform this operation.

petebacondarwin replied:
@alxhub do you have any ideas on how we can remove it though? The trouble is that in the translator the node is hidden in a `WrappedNodeExpr`...

crisbeto replied:
@petebacondarwin @alxhub I've been looking at removing the `createFunctionExpression` call from here, but it'll involve updating the types all across the compiler and potentially losing some type safety. The problem is that this function (`reflectObjectLiteral`) returns a `Map<string, Expression>`, but a `MethodDeclaration` doesn't fit into the `Expression` interface so we can't assign it. The type would have to updated to `Map<string, Node>` which is less strict and it'll involve changing a lot of code since `reflectObjectLiteral` is used all over the compiler.

I see a couple of alternatives:
1. The return value can be changed to `Map<string, Expression|MethodDeclaration>`, but we'll have to add casts to all the places where we currently expect an `Expression` (about 25 places).
2. We can have a function similar to `reflectObjectLiteral` that returns `Map<string, Expression|MethodDeclaration>` and is only used in the relevant places, but then we'd have two functions that basically do the same thing.

What do you think? I wanted to double check before making all these extra changes.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -65,7 +65,10 @@ export class TypeScriptAstFactory implements AstFactory<ts.Statement, ts.Express
 
   createCallExpression(callee: ts.Expression, args: ts.Expression[], pure: boolean): ts.Expression {
-    const call = ts.createCall(callee, undefined, args);
+    const expression = (ts.isFunctionExpression(callee) || ts.isArrowFunction(callee)) ?
+        ts.createParen(callee) :
+        callee;
+    const call = ts.createCall(expression, undefined, args);
```
I really don't like doing this parenthesis wrapping here. In general this is done in the translator. E.g. in [`visitWriteVarExpr`](https://github.com/angular/angular/blob/3ba97ab3917b682b6d08b9e71599051e77e2d3ac/packages/compiler-cli/src/ngtsc/translator/src/translator.ts#L131-L132). 

But I can see that at the point when we are dealing with this call, in the translator, we don't know what this node is - it is a `WrappedNodeExpr`, which is opaque to the translator.

So I don't know if there is a better solution.

If we go with this then there should be an appropriate test added in [typescript_ast_factory_spec.ts](https://github.com/angular/angular/blob/3082f7378b9fd620e11908e4cb0924681f9dea3b/packages/compiler-cli/src/ngtsc/translator/test/typescript_ast_factory_spec.ts#L69-L81).

crisbeto replied:
This logic used to be in the translator, but I had to move it out after the changes that made it generic and I couldn't find a way to infer the node type.

petebacondarwin reviewed (changes requested):
I have a few concerns that are probably not solvable. Perhaps @alxhub or @JoostK have some thoughts?

Apart from that we should definitely change `this: never` to `this: void` throughout.

crisbeto commented on packages/core/src/di/util.ts:
```diff
@@ -35,9 +35,10 @@ export function convertInjectableProviderToFactory(
     return () => ɵɵinject(resolveForwardRef(existingProvider.useExisting));
-  } else if ((provider as FactorySansProvider).useFactory) {
+  } else if ((provider as FactorySansProvider).useFactory!) {
     const factoryProvider = (provider as FactorySansProvider);
-    return () => factoryProvider.useFactory(...injectArgs(factoryProvider.deps || EMPTY_ARRAY));
+    return () => (factoryProvider as any)
+                     .useFactory(...injectArgs(factoryProvider.deps || EMPTY_ARRAY));
```
I'd rather avoid adding type cast functions here since it's performance-sensitive and we won't get much out of the functions, apart from remove a couple of casts.

petebacondarwin replied:
I believe that V8 will inline the function and there will be no performance difference.

crisbeto replied:
To me it feels weird to be depending on V8 internals to optimize something that we only put in place to avoid a type cast during compilation. Since your suggestion about using `void` instead of `never` cleaned up the `as any` casts, I'll keep it as it was before.

petebacondarwin replied:
We rely upon V8 optimisations throughout the Angular code. Consider the efforts we employ to ensure that functions are monomorphic.

Given that the engine will indeed ensure that this is not a problem, it seems strange (conversely) to try to write optimisations into the source (e.g avoiding function calls) rather than focusing on type-safety and IMO clarity of the meaning of the code.

alxhub replied:
The `isFactoryProvider` type guard function is very idiomatic in the compiler, and is commonly used in TypeScript itself (e.g. the `ts.is*` functions which perform assertions on the `kind` of a `ts.Node`). The cost of these monomorphic function calls is extremely minimal even before inlining, and is almost nonexistent afterwards - I wouldn't worry about the overhead.

crisbeto commented on packages/core/src/di/injector.ts:
```diff
@@ -217,7 +217,7 @@ function resolveProvider(provider: SupportedProvider): Record {
     // We need to use USE_VALUE in provider since provider.useValue could be defined as undefined.
     value = (provider as ValueProvider).useValue;
-  } else if ((provider as FactoryProvider).useFactory) {
+  } else if ((provider as FactoryProvider).useFactory!) {
```
The non-null assertion here is necessary, because `strictNullChecks` flags uncalled functions inside `if` statements as of a few TS versions ago.

petebacondarwin replied:
But it is not necessary if you put this inside a type-guard.

crisbeto committed:
fixup! fix(compiler-cli): ignoring useFactory as method declaration in Injectable

crisbeto reviewed (commented):
I've addressed the latest set of feedback on the runtime code. I'm still not sure what to do about the compiler changes.

alxhub reviewed (changes requested):
Pending removal of the synthetic node creation from `ReflectionHost`.

IgorMinar commented on goldens/public-api/core/core.d.ts:
```diff
@@ -337,7 +337,7 @@ export declare interface FactoryProvider extends FactorySansProvider {
 export declare interface FactorySansProvider {
     deps?: any[];
-    useFactory: Function;
+    useFactory: (this: void, ...args: any[]) => any;
```
Isn't this a breaking change? `Function` doesn't seem to be assignable to this new type: https://www.typescriptlang.org/play?#code/DYUwLgBAZgXBBiBXAdgYzASwPbIgXmhXW2QAooss5lEBbAIxACcBKCAbwifESdwqwBqQRAC+AbgBQoSFFgRSYABYYAznABuWDABMANBAB0xgIZMA5uognkATwDaAXTZ4AfNbv5CaTDnKVqOkZWDi4ePmhKYTEpSTkvKHEgA

crisbeto replied:
You're right that it technically could break, but I think that it's unlikely. The signature change wasn't isn't strictly necessary, but we agreed to do it because it prevents users from writing code that might break after compilation. We could roll back the signature change and leave it for the next major version, if we find that it's too break-ey.

IgorMinar reviewed (changes requested):
This issue looks like a good one to fix, but I'm concerned that this might be a breaking change for anyone that currently has factories typed as `Function`. Let's get a better understanding of the impact of this change before we proceed.

I kicked off a presubmit as well.

Reviewed-for: public-api
